File: Patterns/ApnaCollegeDSA/Arrays/Question_1/MajorityElement.java
package Patterns.ApnaCollegeDSA.Arrays.Question_1;

import java.util.*;

public class MajorityElement {
    public static void main(String[] args) {
        Solution s = new Solution();
        // Boyer-Moore voting algo
        s.majorityElement(new int[] { 3, 2, 3 });
        // using hashmap
        int ans = s.majorityElementUsingHashmaps(new int[] { 3, 2, 3 });
        System.out.println(ans);
    }
}

class Solution {
    // Boyer-Moore voting algo
    public int majorityElement(int[] nums) {
        int freq = 0;
        int ans = 0;
        for (int n : nums) {
            if (freq == 0) {
                ans = n;
            }
            if (ans == n) {
                freq++;
            } else {
                freq--;
            }
        }
        return ans;
    }

    // using hashmap
    int majorityElementUsingHashmaps(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i : nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        int max_value = Integer.MIN_VALUE;
        int max_key = Integer.MIN_VALUE;
        for (Map.Entry<Integer, Integer> entries : map.entrySet()) {
            if (entries.getValue() > max_value) {
                max_value = entries.getValue();
                max_key = entries.getKey();
            }
        }
        return max_key; // 3
    }
}

File: Patterns/ApnaCollegeDSA/Arrays/Question_1/MajorityElement.py
class Solution:
    def majorityElement(self, nums):
        map = {}
        for i in nums:
            if map.__contains__(i):
                map[i] += 1
            else:
                map[i] = 1
        max_value = 0
        max_key = 0
        for key,value in map.items():
            if max_value < value:
                max_value = value
                max_key = key
        return max_key
    
sol = Solution()
ans = sol.majorityElement([3,2,3])
print(ans)